OSDOCS-19890 created zero-trust assembly/modules for multi-cluster (#112274)
openshift/openshift-docs: modules/zero-trust-manager-deploy-istio-cr.adoc

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-istio-cr_{context}"]

> = Deploy the Istio custom resource with the federation configuration

Reviewer: ```suggestion
= Deploying the Istio custom resource with the federation configuration
```

[role="_abstract"]
Deploy the Istio custom resource on Cluster A and Cluster B with SPIRE federation and multi-cluster {SMProductName} settings. This configures Istiod to obtain workload certificates from SPIRE and to trust both clusters' SPIFFE bundles for cross-cluster mTLS.

> The Istio CR must include:

Reviewer: Use complete sentence
```suggestion
The Istio CR must include the following fields/parameters/values/options/settings:
```

* `meshConfig.trustDomain` matching the SPIRE trust domain
* `meshConfig.caCertificates` with bundle URLs for both clusters. This handles cross-trust-domain validation.
* `WORKLOAD_IDENTITY_SOCKET_FILE` for SPIRE SDS integration
* `jwksResolverExtraRootCA` for OIDC validation
* Multi-cluster settings which include `meshID`, `clusterName`, and `network`.
* SPIRE injection templates

[NOTE]
====
Do not use `meshConfig.trustDomainAliases`. Use `meshConfig.caCertificates` with `spiffeBundleUrl` instead.
====

.Prerequisites

> * You have completed the "Deploy Istio CNI on both clusters" steps.

Reviewer: similar comment here. Action + For more info..

> * The environment variables from the deploy SPIRE procedure are still set in your shell session.

Reviewer: similar comment as in East West

* The `IstioCNI` DaemonSet is ready on Cluster A and Cluster B.

.Procedure

. Extract the OpenID Connect (OIDC) certificates by running the following commands:
+
[source,terminal]
----
$ export EXTRA_ROOT_CA_A="$(oc get secret oidc-serving-cert \
  --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n ${ZTWIM_NS} -o json | \
  jq -r '.data."tls.crt"' | base64 -d | sed 's/^/        /')"
----
+
[source,terminal]
----
$ export EXTRA_ROOT_CA_B="$(oc get secret oidc-serving-cert \
  --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n ${ZTWIM_NS} -o json | \

>   jq -r '.data."tls.crt"' | base64 -d | sed 's/^/        /')"

Reviewer: split into 2 steps

----

. Get the bundle endpoint URLs by running the following command:
+
[source,terminal]
----
$ export BUNDLE_URL_A="${FEDERATION_ENDPOINT_A}"
----
+
[source,terminal]
----

> $ export BUNDLE_URL_B="${FEDERATION_ENDPOINT_B}"

Reviewer: split into 2 steps

----

. Create the Istio CR on Cluster A by running the following command:
+
[source,terminal]
----
$ oc new-project "${OSSM_NS}" --kubeconfig="${CLUSTER_A_KUBECONFIG}" 2>/dev/null || true

> $ cat <<EOF | oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f -

Reviewer: split into 2 steps

      network: ${NETWORK_A}
    pilot:
      jwksResolverExtraRootCA: |
${EXTRA_ROOT_CA_A}
      env:
        ENABLE_CA_SERVER: "true"
    sidecarInjectorWebhook:
      templates:
        spire: |
          spec:
            initContainers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
        spireGw: |
          spec:
            containers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
EOF
----

. Create the Istio CR on Cluster B by running the following command:
+
[source,terminal]
----
$ oc new-project "${OSSM_NS}" --kubeconfig="${CLUSTER_B_KUBECONFIG}" 2>/dev/null || true

> $ cat <<EOF | oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f -

Reviewer: split into 2 steps

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-istio-cr_{context}"]
= Deploying the Istio custom resource with the federation configuration

[role="_abstract"]

> Deploy the Istio custom resource on Cluster A and Cluster B with SPIRE federation and multi-cluster {SMProductName} settings. This configures Istiod to obtain workload certificates from SPIRE and to trust both clusters' SPIFFE bundles for cross-cluster mTLS.

Reviewer 2: ```suggestion
Deploy the Istio custom resource on Cluster A and Cluster B with SPIRE federation and multi-cluster {SMProductName} settings. This configures Istiod to obtain workload certificates from SPIRE and to trust both of the cluster SPIFFE bundles for cross-cluster mTLS.
```

> The Istio CR must include the following fields and values:

Reviewer 2: ```suggestion
The `Istio` CR must include the following fields and values:
```

> * `meshConfig.trustDomain` matching the SPIRE trust domain

Reviewer 2: ```suggestion
* A `meshConfig.trustDomain` value that matches the SPIRE trust domain.
```
this fragment doesn't make sense to me; is this what it means?

* `meshConfig.caCertificates` with bundle URLs for both clusters. This handles cross-trust-domain validation.
* `WORKLOAD_IDENTITY_SOCKET_FILE` for SPIRE SDS integration
* `jwksResolverExtraRootCA` for OIDC validation
* Multi-cluster settings which include `meshID`, `clusterName`, and `network`.

> * SPIRE injection templates

Reviewer 2: I don't like the mixed sentence and fragment list here; much nicer to use all full sentences with periods if you have to use some.

[NOTE]
====
Do not use `meshConfig.trustDomainAliases`. Use `meshConfig.caCertificates` with `spiffeBundleUrl` instead.
====

.Prerequisites

> * You have completed deploying the Istio CNI on both clusters. For more information, see "Deploy Istio CNI on both clusters" steps.

Reviewer 2: ```suggestion
* You have completed deploying the Istio Container Network Interface (CNI) on both clusters. For more information, see the "Deploy Istio CNI on both clusters" steps.
```

* The environment variables from the SPIRE deployment are set in your shell session.

> * The `IstioCNI` DaemonSet is ready on Cluster A and Cluster B.

Reviewer 2: did we give this verification step earlier?

.Procedure

. Extract the OpenID Connect (OIDC) certificate on Cluster A by running the following command:
+
[source,terminal]
----
$ export EXTRA_ROOT_CA_A="$(oc get secret oidc-serving-cert \
  --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n ${ZTWIM_NS} -o json | \
  jq -r '.data."tls.crt"' | base64 -d | sed 's/^/        /')"
----

. Extract the OpenID Connect (OIDC) certificate on Cluster B by running the following command:
+
[source,terminal]
----
$ export EXTRA_ROOT_CA_B="$(oc get secret oidc-serving-cert \
  --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n ${ZTWIM_NS} -o json | \
  jq -r '.data."tls.crt"' | base64 -d | sed 's/^/        /')"
----

. Get the bundle endpoint URL for Cluster A by running the following command:
+
[source,terminal]
----
$ export BUNDLE_URL_A="${FEDERATION_ENDPOINT_A}"
----

> .. Get the bundle endpoint URL for Cluster B by running the following command:

Reviewer 2: ```suggestion
. Get the bundle endpoint URL for Cluster B by running the following command:
```

+
[source,terminal]
----
$ export BUNDLE_URL_B="${FEDERATION_ENDPOINT_B}"
----

> . Create the Istio CR on Cluster A by running the following command:

Reviewer 2: ```suggestion
. Create the `Istio` custom resource (CR) on Cluster A by running the following command:
```

+
[source,terminal]
----
$ oc new-project "${OSSM_NS}" --kubeconfig="${CLUSTER_A_KUBECONFIG}" 2>/dev/null || true
----

> . Apply the `Istio` custom resource on Cluster A by running the following command:

Reviewer 2: ```suggestion
. Apply the `Istio` CR on Cluster A by running the following command:
```
I see Shubha is also commenting on these object creation/apply steps, so separate all of these. Something like this, https://github.com/openshift/openshift-docs/pull/115220/changes#diff-ee2dfeb79ccd504dc001f3f3abe9a9e0e75c6a85b7e322a9e05f165ff699fb71R42 (you can even skip naming the CR).

        - spiffeBundleUrl: ${BUNDLE_URL_B}
          trustDomains:
            - ${CLUSTER_B_TRUST_DOMAIN}
    global:
      meshID: mesh1
      multiCluster:
        clusterName: ${CLUSTER_A}
      network: ${NETWORK_A}
    pilot:
      jwksResolverExtraRootCA: |
${EXTRA_ROOT_CA_A}
      env:
        ENABLE_CA_SERVER: "true"
    sidecarInjectorWebhook:
      templates:
        spire: |
          spec:
            initContainers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
        spireGw: |
          spec:
            containers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
EOF
----

> . Create the Istio CR on Cluster B by running the following command:

Reviewer 2: ```suggestion
. Create the `Istio` CR on Cluster B by running the following command:
```
remember to backtick when talking about a specific object

+
[source,terminal]
----
$ oc new-project "${OSSM_NS}" --kubeconfig="${CLUSTER_B_KUBECONFIG}" 2>/dev/null || true
----

> . Apply the `Istio` custom resource on Cluster B by running the following command:

Reviewer 2: ```suggestion
. Apply the `Istio` CR on Cluster B by running the following command:
```

        - spiffeBundleUrl: ${BUNDLE_URL_A}
          trustDomains:
            - ${CLUSTER_A_TRUST_DOMAIN}
    global:
      meshID: mesh1
      multiCluster:
        clusterName: ${CLUSTER_B}
      network: ${NETWORK_B}
    pilot:
      jwksResolverExtraRootCA: |
${EXTRA_ROOT_CA_B}
      env:
        ENABLE_CA_SERVER: "true"
    sidecarInjectorWebhook:
      templates:
        spire: |
          spec:
            initContainers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
        spireGw: |
          spec:
            containers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
EOF
----

> . Wait for Istiod to become ready on Cluster A by running the following command:

Reviewer 2: ```suggestion
. Wait for `Istiod` to become ready on Cluster A by running the following command:
```
backtick processes

+
[source,terminal,subs="+quotes,verbatim"]
----
$ until oc get deployment istiod --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n "${OSSM_NS}" &> /dev/null; do
    sleep 3
  done
  oc wait --for=condition=Available deployment/istiod \
    --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n "${OSSM_NS}" --timeout=300s
done
----

> . Wait for Istiod to become ready on Cluster B by running the following command:

Reviewer 2: ```suggestion
. Wait for `Istiod` to become ready on Cluster B by running the following command:
```

        ENABLE_CA_SERVER: "true"
    sidecarInjectorWebhook:
      templates:
        spire: |
          spec:
            initContainers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
        spireGw: |
          spec:
            containers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
----

... Apply the YAML file on Cluster B by running the following command:
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f <filename>
----

. Wait for the `istiod` deployment to be created on Cluster A by running the following command:
+
[source,terminal,subs="+quotes,verbatim"]

> ----
> $ until oc get deployment istiod --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n "${OSSM_NS}" &> /dev/null; do
>   sleep 3
> done
> ----

Reviewer 3: Is this two commands?

```suggestion
----
$ until oc get deployment istiod --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n "${OSSM_NS}" &> /dev/null; do
  sleep 3
----
+
[source,terminal]
----
done
----
```
